File: 883.projection-area-of-3-d-shapes.cpp
/*
 * @lc app=leetcode id=883 lang=cpp
 *
 * [883] Projection Area of 3D Shapes
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  int projectionArea(vector<vector<int>>& grid) {
    return calcXY(grid) + calcXZ(grid) + calcYZ(grid);
  }

  int calcXY(vector<vector<int>>& grid) {
    int size = grid.size();
    int ret = 0;
    for (int y = 0; y < size; y++) {
      for (int x = 0; x < size; x++) {
        if (grid[y][x]) {
          ++ret;
        }
      }
    }
    return ret;
  }

  int calcXZ(vector<vector<int>>& grid) {
    int size = grid.size();
    int ret = 0;
    for (int x = 0; x < size; x++) {
      int max = 0;
      for (int y = 0; y < size; y++) {
        max = std::max(max, grid[y][x]);
      }
      ret += max;
    }
    return ret;
  }

  int calcYZ(vector<vector<int>>& grid) {
    int size = grid.size();
    int ret = 0;
    for (int y = 0; y < size; y++) {
      int max = 0;
      for (int x = 0; x < size; x++) {
        max = std::max(max, grid[y][x]);
      }
      ret += max;
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<vector<int>> grid;

  // Example 1
  grid = {{1, 2}, {3, 4}};
  ASSERT_EQ(s.calcXY(grid), 4);
  ASSERT_EQ(s.calcXZ(grid), 7);
  ASSERT_EQ(s.calcYZ(grid), 6);
  ASSERT_EQ(s.projectionArea(grid), 17);

  // Example 2
  grid = {{2}};
  ASSERT_EQ(s.calcXY(grid), 1);
  ASSERT_EQ(s.calcXZ(grid), 2);
  ASSERT_EQ(s.calcYZ(grid), 2);
  ASSERT_EQ(s.projectionArea(grid), 5);

  // Example 3
  grid = {{1, 0}, {0, 2}};
  ASSERT_EQ(s.calcXY(grid), 2);
  ASSERT_EQ(s.calcXZ(grid), 3);
  ASSERT_EQ(s.calcYZ(grid), 3);
  ASSERT_EQ(s.projectionArea(grid), 8);

  // Example 4
  grid = {{1, 1, 1}, {1, 0, 1}, {1, 1, 1}};
  ASSERT_EQ(s.calcXY(grid), 8);
  ASSERT_EQ(s.calcXZ(grid), 3);
  ASSERT_EQ(s.calcYZ(grid), 3);
  ASSERT_EQ(s.projectionArea(grid), 14);

  // Example 5
  grid = {{2, 2, 2}, {2, 1, 2}, {2, 2, 2}};
  ASSERT_EQ(s.calcXY(grid), 9);
  ASSERT_EQ(s.calcXZ(grid), 6);
  ASSERT_EQ(s.calcYZ(grid), 6);
  ASSERT_EQ(s.projectionArea(grid), 21);
  return 0;
}

File: 883.projection-area-of-3-d-shapes.rs
/*
 * @lc app=leetcode id=883 lang=rust
 *
 * [883] Projection Area of 3D Shapes
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn projection_area(grid: Vec<Vec<i32>>) -> i32 {
        let mut res = 0;
        let mut max_in_cols = vec![0; grid.len()];
        for row in &grid {
            let mut max_in_row = 0;
            for (col, &val) in row.iter().enumerate().filter(|(_, val)| **val > 0) {
                max_in_row = max_in_row.max(val);
                max_in_cols[col] = max_in_cols[col].max(val);
                res += 1;
            }
            res += max_in_row;
        }
        res + max_in_cols.iter().sum::<i32>()
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::projection_area(vec![vec![1, 2], vec![3, 4]]), 17);
    assert_eq!(Solution::projection_area(vec![vec![2]]), 5);
    assert_eq!(Solution::projection_area(vec![vec![1, 0], vec![0, 2]]), 8);
}
